### ISSUE TITLE
prototypes sneakily break :lvalue subs

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9240]
Migrated from [rt.perl.org#51408](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=51408) (status was 'resolved')

Searchable as RT51408$

**Comment by p5pRT** [comment]
### From @mauke

#### Created by @mauke



Hello\!

First\, an unrelated \(but unexpected\, to me\) observation&#8203;:
Function return values don't have the read\-only bit set\, so \.\.\.

&nbsp;  sub f \{ 42 \}
&nbsp;  sub g \{ $\_\[0\] = 1 \}
&nbsp;  g\(f\(\)\);

\.\.\. compiles and runs just fine\, with no visible effect\.

My real problem is related to lvalue subroutines and prototypes\.
Consider the following code&#8203;:

&nbsp;  sub f :lvalue \{ $\_ \}

With this\, I'd expect to be able to use f\(\) anywhere I can use $\_\,
modulo weird things like indirect object syntax\. But code like \.\.\.

&nbsp;  my $ref = \\f\(\);
&nbsp;  $$ref = 1;

\.\.\. works and sets $\_ to 1\. Let's combine this with pass\-by\-alias&#8203;:

&nbsp;  sub f :lvalue \{ $\_ \}
&nbsp;  sub g \{ $\_\[0\] = 1 \}
&nbsp;  g\(f\(\)\);

This code binds g's $\_\[0\] to $\_ and sets $\_ to 1\, as expected\.
However&#8203;:

&nbsp;  sub f :lvalue \{ $\_ \}
&nbsp;  sub g \($\) \{ $\_\[0\] = 1 \}
&nbsp;  g\(f\(\)\);

This code does nothing\. The \($\) prototype on g causes $\_ to be
passed by \(temporary\) value\, and \(per our initial observation\)
this kind of thing silently fails\. g\($\_\) works as expected\, though\,
so there's something weird going on\.
Apparently the perl compiler simply inserts scalar\(\)s in calls to
g\. This works fine for simple variables but breaks lvalue subs\.

Actually\,

&nbsp;  $\{\\$\_\} = 1;   \# ok
&nbsp;  $\{\\f\(\)\} = 1;  \# ok

&nbsp;  $\{\\scalar\($\_\)\} = 1;  \# ok
&nbsp;  $\{\\scalar f\(\)\} = 1;  \# no effect

I think the inconsistency in the last case is a bug in how lvalue
subs work\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl v5.8.8 - Wed Mar 22 20:53:12 CET 2006
It is being executed now by  Perl v5.10.0 - Wed Dec 19 19:01:35 CET 2007.

Site configuration information for perl v5.10.0:

Configured by mauke at Wed Dec 19 19:01:35 CET 2007.

Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.6.22-gentoo-r9, archname=i686-linux
    uname='linux nora 2.6.22-gentoo-r9 #3 preempt sat nov 3 22:12:09 cet 2007 i686 amd athlon(tm) 64 processor 3200+ authenticamd gnulinux '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2 -march=native -fomit-frame-pointer',
    cppflags='-fno-strict-aliasing -pipe -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='4.2.2', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.6.1.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.6.1'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -march=native -fomit-frame-pointer -L/usr/local/lib'

Locally applied patches:
    SAFEARGV0 - disable magic open in <ARGV>
    METHQUAL0 - allow qualified method calls without a valid object
    UNWARN0 - disable stupid warnings for print and qw


@INC for perl v5.10.0:
    /home/mauke/usr/local/lib/perl5/5.10.0/i686-linux
    /home/mauke/usr/local/lib/perl5/5.10.0
    /home/mauke/usr/local/lib/perl5/site_perl/5.10.0/i686-linux
    /home/mauke/usr/local/lib/perl5/site_perl/5.10.0
    .


Environment for perl v5.10.0:
    HOME=/home/mauke
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=POSIX
    LD_LIBRARY_PATH=/home/mauke/usr/local/lib
    LOGDIR (unset)
    PATH=/home/mauke/usr/local/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/i686-pc-linux-gnu/gcc-bin/4.1.2:/opt/sun-jdk-1.4.2.13/bin:/opt/sun-jdk-1.4.2.13/jre/bin:/opt/sun-jdk-1.4.2.13/jre/javaws:/usr/kde/3.5/bin:/usr/qt/3/bin:/usr/games/bin
    PERL_BADLANG (unset)
    PERL_UNICODE=SAL
    SHELL=/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

2008/3/4 via RT l\. mai @&#8203; web\. de \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> My real problem is related to lvalue subroutines and prototypes\.
> Consider the following code&#8203;:
>
>  sub f :lvalue \{ $\_ \}
>
> With this\, I'd expect to be able to use f\(\) anywhere I can use $\_\,
> modulo weird things like indirect object syntax\. But code like \.\.\.
>
>  my $ref = \\f\(\);
>  $$ref = 1;
>
> \.\.\. works and sets $\_ to 1\. Let's combine this with pass\-by\-alias&#8203;:
>
>  sub f :lvalue \{ $\_ \}
>  sub g \{ $\_\[0\] = 1 \}
>  g\(f\(\)\);
>
> This code binds g's $\_\[0\] to $\_ and sets $\_ to 1\, as expected\.
> However&#8203;:
>
>  sub f :lvalue \{ $\_ \}
>  sub g \($\) \{ $\_\[0\] = 1 \}
>  g\(f\(\)\);
>
> This code does nothing\. The \($\) prototype on g causes $\_ to be
> passed by \(temporary\) value\, and \(per our initial observation\)
> this kind of thing silently fails\. g\($\_\) works as expected\, though\,
> so there's something weird going on\.
> Apparently the perl compiler simply inserts scalar\(\)s in calls to
> g\. This works fine for simple variables but breaks lvalue subs\.

It doesn't insert scalar\(\) calls\, but indeed calls f in scalar context
\(instead of list context as in the unprototyped case\)\.

$ diff \-u \<\(bleadperl \-MO=Concise \-e 'sub g;g\(f\(\)\)' 2>&1\) \<\(bleadperl
\-MO=Concise \-e 'sub g\($\);g\(f\(\)\)' 2>&1\)
<details><summary>Inline Patch</summary>

```diff
--- /dev/fd/63	2008-05-28 10:26:06.651464655 +0200
+++ /dev/fd/62	2008-05-28 10:26:06.653628112 +0200
@@ -5,8 +5,8 @@
 8     <1> entersub[t2] vKS/TARG,1 ->9
 -        <1> ex-list K ->8
 3           <0> pushmark s ->4
-6           <1> entersub[t1] lKMS/NO(),TARG,INARGS,1 ->7
--              <1> ex-list lK ->6
+6           <1> entersub[t1] sKMS/NO(),TARG,INARGS,1 ->7
+-              <1> ex-list sK ->6
 4                 <0> pushmark s ->5
 -                 <1> ex-rv2cv sK ->-
 5                    <$> gv(*f) s/EARLYCV ->6


```
</details>
However, lvalue subs work in some scalar contexts:

$ bleadperl \-MO=Concise \-E 'sub f&#8203;:lvalue\{$x\};$y=\(f\(\)=2\);say$x'
e  \<@&#8203;> leave\[1 ref\] vKP/REFC \->\(end\)
1     \<0> enter \->2
2     \<;> nextstate\(main 47 \-e&#8203;:1\) v&#8203;:%\,\{ \->3
9     \<2> sassign vKS/2 \->a
7        \<2> sassign sKPS/2 \->8
3           \<$> const\(IV 2\) s \->4
6           \<1> entersub\[t1\] sKRMS\*/NO\(\)\,TARG\,1 \->7
\-              \<1> ex\-list sK \->6
4                 \<0> pushmark s \->5
\-                 \<1> ex\-rv2cv sK \->\-
5                    \<$> gv\(\*f\) s \->6
\-        \<1> ex\-rv2sv sKRM\*/1 \->9
8           \<$> gvsv\(\*y\) s \->9
a     \<;> nextstate\(main 47 \-e&#8203;:1\) v&#8203;:%\,\{ \->b
d     \<@&#8203;> say vK \->e
b        \<0> pushmark s \->c
\-        \<1> ex\-rv2sv sK/1 \->d
c           \<$> gvsv\(\*x\) s \->d
\-e syntax OK
$ bleadperl \-E 'sub f&#8203;:lvalue\{$x\};$y=\(f\(\)=2\);say$x'
2

I wonder if it's a simple question of fiddling with the entersub flags\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

> Hello\!
> 
> First\, an unrelated \(but unexpected\, to me\) observation&#8203;:
> Function return values don't have the read\-only bit set\, so \.\.\.
> 
>   sub f \{ 42 \}
>   sub g \{ $\_\[0\] = 1 \}
>   g\(f\(\)\);
> 
> \.\.\. compiles and runs just fine\, with no visible effect\.
> 
> My real problem is related to lvalue subroutines and prototypes\.
> Consider the following code&#8203;:
> 
>   sub f :lvalue \{ $\_ \}
> 
> With this\, I'd expect to be able to use f\(\) anywhere I can use $\_\,
> modulo weird things like indirect object syntax\. But code like \.\.\.
> 
>   my $ref = \\f\(\);
>   $$ref = 1;
> 
> \.\.\. works and sets $\_ to 1\. Let's combine this with pass\-by\-alias&#8203;:
> 
>   sub f :lvalue \{ $\_ \}
>   sub g \{ $\_\[0\] = 1 \}
>   g\(f\(\)\);
> 
> This code binds g's $\_\[0\] to $\_ and sets $\_ to 1\, as expected\.
> However&#8203;:
> 
>   sub f :lvalue \{ $\_ \}
>   sub g \($\) \{ $\_\[0\] = 1 \}
>   g\(f\(\)\);
> 
> This code does nothing\. The \($\) prototype on g causes $\_ to be
> passed by \(temporary\) value\, and \(per our initial observation\)
> this kind of thing silently fails\. g\($\_\) works as expected\, though\,
> so there's something weird going on\.
> Apparently the perl compiler simply inserts scalar\(\)s in calls to
> g\. This works fine for simple variables but breaks lvalue subs\.
> 
> Actually\,
> 
>   $\{\\$\_\} = 1;   \# ok
>   $\{\\f\(\)\} = 1;  \# ok
> 
>   $\{\\scalar\($\_\)\} = 1;  \# ok
>   $\{\\scalar f\(\)\} = 1;  \# no effect
> 
> I think the inconsistency in the last case is a bug in how lvalue
> subs work\.

This is now fixed with commit bf8fb5ebdd4\, but I put the wrong bug
number in the commit message\. Oops\!

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
